### PR TITLE
Added notification to OsList instead of OsResults

### PR DESCRIPTION
**Pull request by beeender**
- The notification token for RealmList<RealmModel> will be added to the
  underlying OsList instead of the OsResults created from the list. This
  will have a different behavior for a corner case,
  see https://github.com/realm/realm-object-store/pull/602 . Thus the
  test cases needs to be modified accordingly.
- This would be the precondition of using
  https://github.com/realm/realm-object-store/pull/601
- Add toString() for OrderedCollectionChangeSet.

### CODE REVIEW

File: realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java [hunk 1]
```diff
@@ -130,11 +124,30 @@ private void deleteObjects(Realm realm, int... indices) {
...
-            realm.createObject(Dog.class).setAge(index);
+            Dog dog = realm.createObject(Dog.class);
+            dog.setAge(index);
+            if (type == ObservablesType.REALM_LIST) {
+                Owner owner = realm.where(Owner.class).findFirst();
+                assertNotNull(owner);
+                RealmList<Dog> dogs = owner.getDogs();
+                boolean added = false;
+                for (int i = 0; i < dogs.size(); i++) {
+                    if (dog.getAge() <= dogs.get(i).getAge()) {
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java]: This could use a little description I think. There looks like there is some assumptions about the `index` and `dogs.size()` that I don't fully understand.

File: realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java [hunk 2]
```diff
@@ -130,11 +124,30 @@ private void deleteObjects(Realm realm, int... indices) {
...
+                        dogs.add(i, dog);
+                        added = true;
+                        break;
+                    }
+                }
+                if (!added) {
+                    dogs.add(dog);
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java]: Not sure I understand this one either?

File: realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java [hunk 3]
```diff
@@ -147,9 +160,38 @@ private void modifyObjects(Realm realm, int... indices) {
     }
 
     private void moveObjects(Realm realm, int originAge, int newAge) {
-        realm.where(Dog.class).equalTo(Dog.FIELD_AGE, originAge).findFirst().setAge(newAge);
         if (type == ObservablesType.REALM_LIST) {
-            reorderRealmList(realm);
+            RealmList<Dog> dogs = realm.where(Owner.class).findFirst().getDogs();
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java]: Please add a comment about what this method does

File: realm/realm-library/src/main/java/io/realm/RealmList.java [hunk 4]
```diff
@@ -1281,20 +1259,6 @@ private static boolean isClassForRealmModel(Class<?> clazz) {
...
         throw new IllegalArgumentException("Unexpected value class: " + clazz.getName());
     }
-
-    // TODO: Object Store is not able to merge change set for links list. Luckily since we were still using LinkView
-    // when ship the fine grain notifications, the listener on RealmList is actually added to a OS Results which is
-    // created from the link view. OS Results is computing the change set by comparing the old/new collection. So it
-    // will give the right results if you remove all elements from a RealmList then add all them back and add one more
-    // new element. By right results it means the change set only include one insertion. But if the listener is on the
-    // OS List, the change set will include all ranges of th list. So we keep the old behaviour for
-    // RealmList<RealmModel> for now. See https://github.com/realm/realm-object-store/issues/541
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmList.java]: I tend to agree that this sounds like a mistake, but if we introduce the behaviour (until fixed in OS), then it doesn't seem like a major issue either.

File: realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java (hunk 1 above)
Comment (beeender) [reply realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java]: added some comments. The assumption is the both `RealmResults` and `RealmList` are sorted by age. See `populateData()`.

File: realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java (hunk 2 above)
Comment (beeender) [reply realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java]: ditto

File: realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java (hunk 3 above)
Comment (beeender) [reply realm/realm-library/src/androidTest/java/io/realm/OrderedCollectionChangeSetTests.java]: simplified the code and added some comments.
